File: mobile/src/main/java/guepardoapps/mynote/service/FloatingService.java
package guepardoapps.mynote.service;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.graphics.PixelFormat;
import android.os.IBinder;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.WindowManager;
import android.view.WindowManager.LayoutParams;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.ProgressBar;

import com.rey.material.widget.FloatingActionButton;

import guepardoapps.mynote.R;
import guepardoapps.mynote.activities.ActivityAdd;
import guepardoapps.mynote.activities.ActivityImpressum;
import guepardoapps.mynote.activities.ActivitySettings;
import guepardoapps.mynote.common.constants.SharedPrefConstants;
import guepardoapps.mynote.controller.DatabaseController;
import guepardoapps.mynote.controller.NavigationController;
import guepardoapps.mynote.controller.SharedPrefController;
import guepardoapps.mynote.customadapter.NoteListAdapter;

public class FloatingService extends Service {
    private Context _context;
    private NavigationController _navigationController;
    private SharedPrefController _sharedPrefController;

    private WindowManager _bubbleViewManager;
    private ImageView _bubble;
    private LayoutParams _bubbleParamsStore;
    private int _bubblePosY = 100;
    private Boolean _movedBubble = false;

    private WindowManager _listViewManager;
    private View _listView;

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onCreate() {
        super.onCreate();

        _context = this;
        _navigationController = new NavigationController(_context);
        _sharedPrefController = new SharedPrefController(_context, SharedPrefConstants.SHARED_PREF_NAME);

        _bubblePosY = _sharedPrefController.LoadIntegerValueFromSharedPreferences(SharedPrefConstants.BUBBLE_POS_Y);

        _bubbleViewManager = (WindowManager) getSystemService(WINDOW_SERVICE);

        _bubble = new ImageView(_context);
        _bubble.setImageResource(R.mipmap.ic_launcher);

        final LayoutParams params = new LayoutParams(
                LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT,
                LayoutParams.TYPE_PHONE, LayoutParams.FLAG_NOT_FOCUSABLE,
                PixelFormat.TRANSLUCENT);

        params.gravity = Gravity.TOP | Gravity.START;
        params.x = 0;
        params.y = _bubblePosY;

        _bubbleParamsStore = params;

        _bubble.setOnTouchListener(new View.OnTouchListener() {
            private int initialX;
            private int initialY;
            private float initialTouchX;
            private float initialTouchY;

            @Override
            public boolean onTouch(View view, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        if (params.x < 0) {
                            params.x = 0;

                            _bubbleParamsStore = params;
                            _bubbleViewManager.updateViewLayout(_bubble, _bubbleParamsStore);
                        }
                        initialX = params.x;
                        initialY = params.y;
                        initialTouchX = event.getRawX();
                        initialTouchY = event.getRawY();
                        return true;
                    case MotionEvent.ACTION_UP:
                        view.performClick();
                        return true;
                    case MotionEvent.ACTION_MOVE:
                        params.x = initialX + (int) (event.getRawX() - initialTouchX);
                        params.y = initialY + (int) (event.getRawY() - initialTouchY);
                        if (initialX - params.x > 25 || initialY - params.y > 25 || params.x - initialX > 25
                                || params.y - initialY > 25) {
                            _movedBubble = true;
                        }

                        _bubbleParamsStore = params;
                        _bubbleViewManager.updateViewLayout(_bubble, _bubbleParamsStore);
                        return true;
                }
                return false;
            }
        });

        _bubble.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View view) {
                if (_movedBubble) {
                    _movedBubble = false;
                    params.x = 0;
                    _bubblePosY = params.y;

                    _sharedPrefController.SaveIntegerValue(SharedPrefConstants.BUBBLE_POS_Y, _bubblePosY);

                    _bubbleParamsStore = params;
                    _bubbleViewManager.updateViewLayout(_bubble, _bubbleParamsStore);
                } else {
                    showListView();
                }
            }
        });
        _bubbleViewManager.addView(_bubble, _bubbleParamsStore);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (_bubble != null) {
            _bubbleViewManager.removeView(_bubble);
        }
    }

    private void showListView() {
        _listViewManager = (WindowManager) _context.getApplicationContext()
                .getSystemService(Context.WINDOW_SERVICE);
        WindowManager.LayoutParams layoutParams = new WindowManager.LayoutParams();

        layoutParams.gravity = Gravity.CENTER;
        layoutParams.type = WindowManager.LayoutParams.TYPE_SYSTEM_ALERT;
        layoutParams.width = WindowManager.LayoutParams.MATCH_PARENT;
        layoutParams.height = WindowManager.LayoutParams.WRAP_CONTENT;
        layoutParams.alpha = 1.0f;
        layoutParams.packageName = _context.getPackageName();
        layoutParams.buttonBrightness = 1f;
        layoutParams.windowAnimations = android.R.style.Animation_Dialog;

        _listView = View.inflate(_context.getApplicationContext(), R.layout.side_main, null);

        FloatingActionButton btnSettings = _listView.findViewById(R.id.btnSettings);
        btnSettings.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                _navigationController.NavigateTo(ActivitySettings.class, false);
                _listViewManager.removeView(_listView);
            }
        });

        FloatingActionButton btnImpressum = _listView.findViewById(R.id.btnImpressum);
        btnImpressum.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                _navigationController.NavigateTo(ActivityImpressum.class, false);
                _listViewManager.removeView(_listView);
            }
        });

        FloatingActionButton btnAdd = _listView.findViewById(R.id.goToAddView);
        btnAdd.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                _navigationController.NavigateTo(ActivityAdd.class, false);
                _listViewManager.removeView(_listView);
            }
        });

        FloatingActionButton btnClose = _listView.findViewById(R.id.btnClose);
        btnClose.setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View view) {
                _listViewManager.removeView(_listView);
            }
        });

        ListView noteListView = _listView.findViewById(R.id.listView);
        ProgressBar progressBar = _listView.findViewById(R.id.progressBar);

        noteListView.setAdapter(new NoteListAdapter(_context, DatabaseController.getInstance().GetNotes()));

        progressBar.setVisibility(View.GONE);
        noteListView.setVisibility(View.VISIBLE);

        _listViewManager.addView(_listView, layoutParams);
    }
}

File: mobile/src/main/java/guepardoapps/mynote/services/FloatingService.kt
package guepardoapps.mynote.services

import android.annotation.SuppressLint
import android.app.Service
import android.content.Context
import android.content.Intent
import android.graphics.PixelFormat
import android.graphics.drawable.GradientDrawable
import android.os.IBinder
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.view.WindowManager
import android.widget.ImageView
import android.widget.ListView
import android.widget.ProgressBar
import com.rey.material.widget.FloatingActionButton
import guepardoapps.mynote.R
import guepardoapps.mynote.activities.ActivityAbout
import guepardoapps.mynote.activities.ActivityAdd
import guepardoapps.mynote.activities.ActivitySettings
import guepardoapps.mynote.controller.NavigationController
import guepardoapps.mynote.controller.SharedPreferenceController
import guepardoapps.mynote.controller.SystemInfoController
import guepardoapps.mynote.customadapter.NoteListAdapter
import guepardoapps.mynote.database.note.DbNote
import guepardoapps.mynote.utils.Logger

// https://stackoverflow.com/questions/7569937/unable-to-add-window-android-view-viewrootw44da9bc0-permission-denied-for-t#answer-34061521

@ExperimentalUnsignedTypes
class FloatingService : Service() {

    private lateinit var bubbleView: ImageView

    private lateinit var bubbleWindowManager: WindowManager

    private lateinit var listViewWindowManager: WindowManager

    private lateinit var navigationController: NavigationController

    private lateinit var systemInfoController: SystemInfoController

    override fun onBind(intent: Intent?): IBinder? = null

    override fun onCreate() {
        super.onCreate()

        bubbleView = ImageView(this)
        bubbleWindowManager = getSystemService(Context.WINDOW_SERVICE) as WindowManager
        listViewWindowManager = getSystemService(Context.WINDOW_SERVICE) as WindowManager
        navigationController = NavigationController(this)
        systemInfoController = SystemInfoController(this)

        initBubbleView()
    }

    override fun onDestroy() {
        super.onDestroy()
        try {
            bubbleWindowManager.removeView(bubbleView)
        } catch (exception: Exception) {
            Logger.instance.error(FloatingService::class.java.simpleName, exception)
        }
    }

    @Suppress("DEPRECATION")
    @SuppressLint("ClickableViewAccessibility")
    private fun initBubbleView() {
        val sharedPreferenceController = SharedPreferenceController(this)

        var bubbleMoved = false
        var bubbleParamsStore: WindowManager.LayoutParams?

        val params = WindowManager.LayoutParams(
                WindowManager.LayoutParams.WRAP_CONTENT,
                WindowManager.LayoutParams.WRAP_CONTENT,
                WindowManager.LayoutParams.TYPE_PHONE,
                WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
                PixelFormat.TRANSLUCENT).apply {
            gravity = Gravity.TOP or Gravity.START
            x = sharedPreferenceController.load(getString(R.string.sharedPrefBubblePosX), resources.getInteger(R.integer.sharedPrefBubbleDefaultPosX))
            y = sharedPreferenceController.load(getString(R.string.sharedPrefBubblePosY), resources.getInteger(R.integer.sharedPrefBubbleDefaultPosY))
            type = WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
        }

        val backgroundShape = GradientDrawable()
        backgroundShape.setColor(resources.getColor(R.color.colorPrimaryDark))
        backgroundShape.cornerRadius = 100.0f

        bubbleView.background = backgroundShape
        bubbleView.setImageResource(R.mipmap.ic_launcher)
        bubbleView.setOnTouchListener(object : View.OnTouchListener {
            private var initialX: Int = 0
            private var initialY: Int = 0
            private var initialTouchX: Float = 0.toFloat()
            private var initialTouchY: Float = 0.toFloat()

            override fun onTouch(view: View, event: MotionEvent): Boolean {
                when (event.action) {
                    MotionEvent.ACTION_DOWN -> {
                        if (params.x < 0) {
                            params.x = 0
                            bubbleParamsStore = params
                            bubbleWindowManager.updateViewLayout(bubbleView, bubbleParamsStore)
                        }

                        initialX = params.x
                        initialY = params.y

                        initialTouchX = event.rawX
                        initialTouchY = event.rawY

                        return true
                    }
                    MotionEvent.ACTION_UP -> {
                        view.performClick()
                        return true
                    }
                    MotionEvent.ACTION_MOVE -> {
                        params.x = initialX + (event.rawX - initialTouchX).toInt()
                        params.y = initialY + (event.rawY - initialTouchY).toInt()

                        val minMovement = resources.getInteger(R.integer.bubbleMinMove)
                        bubbleMoved = initialX - params.x > minMovement
                                || initialY - params.y > minMovement
                                || params.x - initialX > minMovement
                                || params.y - initialY > minMovement

                        bubbleParamsStore = params
                        bubbleWindowManager.updateViewLayout(bubbleView, bubbleParamsStore)

                        return true
                    }
                }

                return false
            }
        })

        bubbleView.setOnClickListener {
            if (bubbleMoved) {
                bubbleMoved = false

                params.x = if (params.x > systemInfoController.displayDimension().width / 2) systemInfoController.displayDimension().width else 0

                sharedPreferenceController.save(getString(R.string.sharedPrefBubblePosX), params.x)
                sharedPreferenceController.save(getString(R.string.sharedPrefBubblePosY), params.y)

                bubbleParamsStore = params
                bubbleWindowManager.updateViewLayout(bubbleView, bubbleParamsStore)
            } else {
                showListView()
            }
        }

        bubbleParamsStore = params
        bubbleWindowManager.addView(bubbleView, bubbleParamsStore)
    }

    private fun showListView() {
        val listView = View.inflate(applicationContext, R.layout.side_main, null)

        listView.findViewById<FloatingActionButton>(R.id.btnSettings)
                .setOnClickListener {
                    navigationController.navigate(ActivitySettings::class.java, false)
                    listViewWindowManager.removeView(listView)
                }

        listView.findViewById<FloatingActionButton>(R.id.btnAbout)
                .setOnClickListener {
                    navigationController.navigate(ActivityAbout::class.java, false)
                    listViewWindowManager.removeView(listView)
                }

        listView.findViewById<FloatingActionButton>(R.id.goToAddView)
                .setOnClickListener {
                    navigationController.navigate(ActivityAdd::class.java, false)
                    listViewWindowManager.removeView(listView)
                }

        listView.findViewById<FloatingActionButton>(R.id.btnClose)
                .setOnClickListener {
                    listViewWindowManager.removeView(listView)
                }

        listView.findViewById<ListView>(R.id.listView).apply {
            adapter = NoteListAdapter(context, DbNote(context).get().toTypedArray()) { showListView() }
            visibility = View.VISIBLE
        }
        listView.findViewById<ProgressBar>(R.id.progressBar).visibility = View.GONE

        listViewWindowManager.addView(listView, WindowManager.LayoutParams()
                .apply {
                    gravity = Gravity.CENTER
                    type = WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
                    width = WindowManager.LayoutParams.MATCH_PARENT
                    height = WindowManager.LayoutParams.WRAP_CONTENT
                    alpha = 1.0f
                    packageName = applicationContext.packageName
                    buttonBrightness = 1f
                    windowAnimations = android.R.style.Animation_Dialog
                })
    }
}